miden-lib/asm/kernels/transaction/api.masm

```diff
@@ -863,20 +953,103 @@ export.get_note_serial_number
+#! - VAULT_ROOT is the commitment of the foreign account's vault.
+#! - STORAGE_ROOT is the commitment of the foreign account's storage.
+#! - STORAGE_SLOT_DATA is the data contained in the storage slot which is constructed as follows: 
+#!   [SLOT_VALUE, slot_type, 0, 0, 0]
+#! - CODE_ROOT is the commitment of the foreign account's code.
+#! - ACCOUNT_PROCEDURE_DATA is the information about account procedure which is constructed as 
+#!   follows: [PROCEDURE_MAST_ROOT, storage_offset, 0, 0, 0]
 #!
 #! Panics if:
 #! - the current context is not a native context.
 export.start_foreign_context
     # check that this procedure was executed against the native account
     exec.memory::assert_native_account
+    # OS => [KERNEL_PROCEDURE_HASH, foreign_account_id]
 
-    dropw drop # unimplemented!
+    # drop the procedure's hash
+    dropw
+    # OS => [foreign_account_id]
+
+    # get the memory address and a flag whether this account was already loaded.
+    exec.account::get_foreign_account_ptr
+    # OS => [was_loaded, ptr, foreign_account_id]
+
+    if.true
+        exec.memory::set_current_account_data_ptr drop
+        # OS => []
+    else
+        exec.memory::set_current_account_data_ptr
+        # OS => [foreign_account_id]
+
+        # construct the word with account ID to load the core account data from the advice map
+        push.0.0.0
+        # OS => [0, 0, 0, foreign_account_id]
+
+        # move the core account data to the advice stack
+        adv.push_mapval
+        # OS => [0, 0, 0, foreign_account_id]
+        # AS => [[foreign_account_id, 0, 0, account_nonce], VAULT_ROOT, STORAGE_ROOT, CODE_ROOT]
+
+        # store the id and nonce of the foreign account to the memory
+        dropw adv_loadw 
+        exec.memory::set_acct_id_and_nonce dropw
+        # OS => []
+        # AS => [VAULT_ROOT, STORAGE_ROOT, CODE_ROOT]
+
+        # store the vault root of the foreign account to the memory
+        adv_loadw exec.memory::set_acct_vault_root dropw
+        # OS => []
+        # AS => [STORAGE_ROOT, CODE_ROOT]
+
+        # move the storage root and the code root to the operand stack 
+        adv_loadw padw adv_loadw
+        # OS => [CODE_ROOT, STORAGE_ROOT]
+        # AS => []
+
+        # store the code root into the memory
+        exec.memory::set_acct_code_commitment
+        # OS => [CODE_ROOT, STORAGE_ROOT]
+        # AS => []
+
+        # save the account procedure data into the memory
+        exec.account::save_account_procedure_data
+        # OS => [STORAGE_ROOT]
+        # AS => []
+
+        # store the storage root to the memory
+        exec.memory::set_acct_storage_commitment
+        # OS => [STORAGE_ROOT]
+        # AS => []
+
+        # save the storage slots data into the memory
+        exec.account::save_account_storage_data
+        # OS => []
+        # AS => []
+    end
+
+    # make sure that the state of the loaded foreign account is valid
+    exec.account::validate_current_foreign_account
```

Comment:
I would expand the comment here a bit to explain that here we are checking account state agains the account database references by the transaction (or something like that).